corehq/apps/es/tests/utils.py

```diff
@@ -124,101 +125,119 @@ def es_test_decorator(test):
     if not requires:
         return es_test_attr(test)
 
-    def setup_func():
-        comment = f"created for {test.__module__}.{test.__qualname__}"
-        for adapter in iter_all_doc_adapters():
-            setup_test_index(adapter, comment)
+    comment = f"created for {test.__module__}.{test.__qualname__}"
+    operations = _index_operations(requires, comment)
+    if isclass(test):
+        test = _add_setup_and_cleanup(test, setup_class, operations)
+    else:
+        if setup_class:
+            raise ValueError(f"keyword 'setup_class' is for class decorators, test={test}")
+        test = _decorate_test_function(test, operations)
+    return es_test_attr(test)
+
 
-    def teardown_func():
-        for adapter in iter_all_doc_adapters():
-            teardown_test_index(adapter)
+def _index_operations(adapters, comment):
 
-    def iter_all_doc_adapters():
+    def iter_all_adapters():
         for adapter in adapters:
             if isinstance(adapter, ElasticMultiplexAdapter):
                 yield adapter.primary
                 yield adapter.secondary
             else:
                 yield adapter
 
-    adapters = list(requires)
-    if isclass(test):
-        test = _add_setup_and_teardown(test, setup_class, setup_func, teardown_func)
-    else:
-        if setup_class:
-            raise ValueError(f"keyword 'setup_class' is for class decorators, test={test}")
-        test = _decorate_test_function(test, setup_func, teardown_func)
-    return es_test_attr(test)
+    operations = {}
+    for adapter in iter_all_adapters():
+        operations[adapter.index_name] = CreateIndex(
+            adapter.index_name,
+            adapter.type,
+            adapter.mapping,
+            adapter.analysis,
+            adapter.settings_key,
+            comment=comment,
+        )
+    return operations
 
 
-def _decorate_test_function(test, setup_func, teardown_func):
+def _decorate_test_function(test, operations):
 
     @wraps(test)
     def wrapper(*args, **kw):
-        setup_func()
+        created = []
         try:
+            for index_name, operation in operations.items():
+                _run_create_index_operation(index_name, operation)
+                created.append(operation)
             return test(*args, **kw)
         finally:
-            teardown_func()
+            for operation in created:
```

Comment:
Should this iterate in reverse so the cleanup happens in the revers order of setup? Maybe it doesn't matter in this case?
```suggestion
            for operation in reversed(created):
```

Reply:
Good point! In this case it doesn't matter (yet, anyhow), but your suggestion is the correct way to do it to prevent gotchas in the event it someday would matter.